enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/CoreStateDownloader.java

```diff
@@ -71,97 +76,160 @@ public CoreStateDownloader( LocalDatabase localDatabase, Lifecycle startStopOnSt
+            if ( catchupResult == E_TRANSACTION_PRUNED )
             {
-                @Override
-                public void onCoreSnapshot( CompletableFuture<CoreSnapshot> signal, CoreSnapshot response )
-                {
-                    signal.complete( response );
-                }
-            } );
-
-            if ( isEmptyStore )
+                log.warn( format( "Failed to pull transactions from (%s). They may have been pruned away", primary ) );
+                localDatabase.delete();
+                isEmptyStore = true;
+            }
+            else if ( catchupResult != SUCCESS_END_OF_STREAM )
+            {
+                log.warn( format( "Unexpected catchup operation result %s from %s", catchupResult, primary ) );
+                return false;
+            }
+        }
+
+        if ( isEmptyStore )
+        {
+            try
             {
                 storeCopyProcess.replaceWithStoreFrom( addressProvider, remoteStoreId );
             }
-            else
+            catch ( StoreCopyFailedException e )
             {
-                StoreId localStoreId = localDatabase.storeId();
-                CatchupResult catchupResult = remoteStore.tryCatchingUp( primary, localStoreId, localDatabase.storeDir(), false );
-
-                if ( catchupResult == E_TRANSACTION_PRUNED )
-                {
-                    log.info( format( "Failed to pull transactions from (%s). They may have been pruned away", primary ) );
-                    localDatabase.delete();
-
-                    storeCopyProcess.replaceWithStoreFrom( addressProvider, localStoreId );
-                }
-                else if ( catchupResult != SUCCESS_END_OF_STREAM )
-                {
-                    throw new StoreCopyFailedException( "Failed to download store: " + catchupResult );
-                }
+                log.warn( "Failed to copy and replace store", e );
+                return false;
             }
-
-            /* We install the snapshot after the store has been downloaded,
-             * so that we are not left with a state ahead of the store. */
-            snapshotService.installSnapshot( coreSnapshot );
-            log.info( "Core snapshot installed: " + coreSnapshot );
-
-            /* Starting the database will invoke the commit process factory in
-             * the EnterpriseCoreEditionModule, which has important side-effects. */
-            log.info( "Starting local database" );
-            localDatabase.start();
-            coreStateMachines.installCommitProcess( localDatabase.getCommitProcess() );
-            startStopOnStoreCopy.start();
         }
-        catch ( StoreCopyFailedException e )
+
+        /* We install the snapshot after the store has been downloaded,
+         * so that we are not left with a state ahead of the store. */
+        snapshotService.installSnapshot( coreSnapshot );
```

Comment:
are we okay with throwing whatever here?

Reply:
It only declares I/O exceptions, and I think such should be thrown and be handled fatally.